File: src/main/java/com/kkllffaa/meteor_litematica_printer/Modules/CRUD/AtomicSettings/CommonSettings.java
package com.kkllffaa.meteor_litematica_printer.Modules.CRUD.AtomicSettings;

import meteordevelopment.meteorclient.systems.modules.Module;
import meteordevelopment.meteorclient.settings.SettingGroup;
import meteordevelopment.meteorclient.settings.Setting;
import meteordevelopment.meteorclient.MeteorClient;
import meteordevelopment.meteorclient.settings.DoubleSetting;
import meteordevelopment.meteorclient.settings.EnumSetting;
import net.minecraft.network.packet.c2s.play.HandSwingC2SPacket;
import net.minecraft.util.Hand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Box;


import com.kkllffaa.meteor_litematica_printer.Addon;
import com.kkllffaa.meteor_litematica_printer.Functions.MyUtils.ActionMode;
import com.kkllffaa.meteor_litematica_printer.Functions.MyUtils.DistanceMode;



public class CommonSettings extends Module {
    public static CommonSettings Instance = new CommonSettings();
    public CommonSettings() {
		super(Addon.SettingsForCRUD, "Common", "Module to configure AtomicSettings.");
        this.toggle();
	}

	@Override
	public void toggle() {
        if (isActive()) {
            return;
        }
        super.toggle();
    }
    
	private final SettingGroup sgGeneral = settings.getDefaultGroup();
    private final Setting<ActionMode> swingHand = sgGeneral.add(new EnumSetting.Builder<ActionMode>()
        .name("swing-hand")
        .description("swing hand post mining.")
        .defaultValue(ActionMode.None)
        .build()
    );
    
    private final Setting<DistanceMode> distanceProtection = sgGeneral.add(new EnumSetting.Builder<DistanceMode>()
        .name("distance-protection")
        .description("Prevent CRUD blocks that are too far to the player.")
        .defaultValue(DistanceMode.Max)
        .build()
    );

    private final Setting<Double> maxDistance = sgGeneral.add(new DoubleSetting.Builder()
        .name("max-distance")
        .description("Maximum distance from player to mine block face.")
        .defaultValue(5.4)
        .min(1.0)
        .max(1024.0)
        .sliderRange(1.0, 5.5)
        .visible(() -> distanceProtection.get() == DistanceMode.Max)
        .build()
    );
    public static void swing(Hand hand) {
        Instance.swingHand(hand);
    }
    private void swingHand(Hand hand) {
        switch (swingHand.get()) {
            case ActionMode.None -> {}
            case ActionMode.SendPacket -> MeteorClient.mc.getNetworkHandler().sendPacket(new HandSwingC2SPacket(hand));
            case ActionMode.Normal -> MeteorClient.mc.player.swingHand(hand);
        }
    }

    public static double getHandDistance() {
        return Instance.getHandDistanceStep();
    }
    private double getHandDistanceStep() {
        return switch(distanceProtection.get()){
            case Auto -> {
                double playerHandRange = mc.player.getBlockInteractionRange();
                yield mc.player.isCreative() ? playerHandRange + 0.5 : playerHandRange;
            }
            case Max -> maxDistance.get();
        };
    }

    public static boolean canTouchTheBlockAt(BlockPos pos) {
        return Instance.canInteractWithBlockAt(pos);
    }

    private boolean canInteractWithBlockAt(BlockPos pos) {
        return switch (distanceProtection.get()) {
            case Auto ->
                mc.player.canInteractWithBlockAt(pos, mc.player.isCreative() ? 0.5 : 0.0);
            case Max ->
                (new Box(pos)).squaredMagnitude(mc.player.getEyePos()) < maxDistance.get() * maxDistance.get();
        };
    }
}

File: src/main/kotlin/com/kkllffaa/meteor_litematica_printer/Modules/CRUD/AtomicSettings/CommonSettings.kt
package com.kkllffaa.meteor_litematica_printer.Modules.CRUD.AtomicSettings

import com.kkllffaa.meteor_litematica_printer.Addon
import com.kkllffaa.meteor_litematica_printer.Functions.*
import meteordevelopment.meteorclient.settings.*
import meteordevelopment.meteorclient.systems.modules.Module
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Box

object CommonSettings : Module(Addon.SettingsForCRUD, "Common", "Module to configure AtomicSettings.") {
    override fun toggle() {
        if (isActive) return
        super.toggle()
    }

    init {
        toggle()
    }

    private val sgGeneral = settings.defaultGroup
    private val sgOther = settings.createGroup("Other")

    private val distanceProtection: Setting<DistanceMode> = sgGeneral.add(
        EnumSetting.Builder<DistanceMode>()
            .name("distance-protection")
            .description("Prevent CRUD blocks that are too far to the player.")
            .defaultValue(DistanceMode.Max)
            .build()
    )

    private val maxDistance: Setting<Double> = sgGeneral.add(
        DoubleSetting.Builder()
            .name("max-distance")
            .description("Maximum distance from player to mine block face.")
            .defaultValue(5.49)
            .min(1.0)
            .max(1024.0)
            .sliderRange(1.0, 5.5)
            .visible { distanceProtection.get() == DistanceMode.Max }
            .build()
    )
    val SignMaxTextWidth: Setting<Int> = sgOther.add(
        IntSetting.Builder()
            .name("Sign-max-text-width")
            .description("0 for default auto")
            .defaultValue(99999)
            .min(0)
            .build()
    )

    val NoFogInLava: Setting<Boolean> = sgOther.add(
        BoolSetting.Builder()
            .name("No-fog-in-lava")
            .description("See blocks in lava.")
            .defaultValue(true)
            .build()
    )

    val allowWhitespace: Setting<Boolean> = sgOther.add(
        BoolSetting.Builder()
            .name("allow-whitespace-in-everything")
            .description("Allow whitespace characters in text everywhere.")
            .defaultValue(true)
            .build()
    )


    val PlayerHandDistance: Double
        get() = when (distanceProtection.get()) {
            DistanceMode.Auto -> {
                mc.player?.let {
                    val playerHandRange = it.blockInteractionRange
                    if (it.isCreative) playerHandRange + 0.5 else playerHandRange
                } ?: 4.5
            }

            DistanceMode.Max -> maxDistance.get()
        }


    fun playerCanTouchBlockPos(pos: BlockPos): Boolean =
        // DistanceMode.Auto -> player.canInteractWithBlockAt(pos, if (player.isCreative) 0.5 else 0.0)
        mc.player?.let {
            PlayerHandDistance.let { maxDist ->
                Box(pos).squaredMagnitude(it.eyePos) < maxDist * maxDist
            }
        } ?: false

}
